doc/user/autotune.rst

```diff
@@ -0,0 +1,173 @@
+Autotuning
+==========
+The kernels shown so far all have a fixed work-group size. However, it's not
+easy to know in advance what the best work-group size is for a specific piece
+of hardware, and even harder to know what it should be in code that may run on
+multiple generations of hardware. Furthermore, work-group size is just the
+most common tuning parameter, but there might be others.
+
+A common approach to this problem is to use "autotuning": benchmark
+different options on the fly and then use the best. There are two variants of
+this approach, which I'll call "on-line" and "off-line". In on-line
+autotuning, each time a kernel is invoked by the user, a different set of
+parameters is tested; the benchmarking is a side effect of doing useful work,
+but the useful work will have variable performance until the tuner has
+converged. In off-line autotuning, benchmarking is done on a synthetic
+workload and the optimal parameters are selected before doing any real work.
+
+On-line autotuning has the advantage that it does not require a long wait
+before useful work can be done, and also that the workload being benchmarked
+will be representative. However, katsdpsigproc was originally developed for
+use in real-time pipelines where highly variable performance during the tuning
+phase is unacceptable. It thus only supports off-line autotuning.
+
+Despite being off-line, the autotuning is transparent and automatic to users
+of operations. If an operation template is constructed in a new configuration,
+the autotuning will be run. The result is saved in a sqlite database so that
+it will not need to be run again. Results are indexed by the device and driver
+version so that old results will not be reused for new hardware for which they
+might not be appropriate.
+
+To take advantage of autotuning, authors of operations will need to add code
+to their operation templates to specify the parameter space to search and the
+code to benchmark. Let's see how that might look for our
+:class:`!MultiplyTemplate` class (see :ref:`operation-templates`), to make it
+automatically determine the work group size.
+
+.. code:: python
+
+    class MultiplyTemplate:
+        def __init__(self, context, tuning=None):
+            if tuning is None:
+                tuning = self.autotune(context)
+            self.wgs = tuning['wgs']
+            self.program = build(context, '', source=SOURCE)
+
+        @classmethod
+        @katsdpsigproc.tune.autotuner(test={'wgs': 32})
```

Comment:
If I try this I get:
Exception has occurred: AttributeError
module 'katsdpsigproc' has no attribute 'tune' - is there something else that is required?

Reply:
Did you `import katsdpsigproc.tune`?